walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,174 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  private List<TimeRange> attendeesEventTimes;
+  private Map<String, List<TimeRange>> optionalAttendeesEventTimes;
+  private List<List<String>> optionalAttendeesCombs;
+
+  /** 
+   * Finds a list of times when the requested event can happen, 
+   * given the list of current events and the request information. 
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendees = new ArrayList<>(request.getAttendees());
+    Collection<String> optionalAttendees = new ArrayList<>(request.getOptionalAttendees());
+    // On the webapp, if form is left empty, an empty string is obtained. Remove them here.
+    attendees.removeAll(Collections.singleton(""));
+    optionalAttendees.removeAll(Collections.singleton(""));
+
+    long duration = request.getDuration();
+
+    // Populate {@code attendeesEventTimes} and {@code optionalAttendeesEventTimes}.
+    getAttendeesEventTimes(events, attendees, optionalAttendees);
+
+    Collection<TimeRange> possibleTimes = new ArrayList<>();
+    if (attendees.isEmpty()) {
+      // If there is no mandatory attendee, only consider availabilities of optional attendees.
+
+      // {@code eventTimes} is the list of current events' time ranges that need to be avoided when scheduling.
```

Comment:
The position of this comment implies that it is for the line below. You don't need to mention the name.
```suggestion
      // A list of current events' time ranges that need to be avoided when scheduling.
```

Reply:
Ok. Would it be good to leave an empty line above to say those are two separate comments?